### PR TITLE
Update NestedTensor docs

### PR DESCRIPTION
**Pull request by cpuhrsch**

### CODE REVIEW

**Review by george-qi — APPROVED** [review]

**Comment by vadimkantorov** [comment]
Should the docs mention that it forces a copy/realloc of the passed tensors?

Another frequent question is on how many dims can be different (I think someone asked on application of this to batch images that all have different spatial sizes)

Also on semantics, along which dimension they are "nested" (I think always batched across dim=0, right?)

**Comment by cpuhrsch** [comment]
@vadimkantorov - I wasn't planning on adding more than just removing the sentence that says "autograd is disabled", but I'll add in your suggestions. Thanks!

File: docs/source/nested.rst [hunk 1]
```diff
@@ -39,6 +39,20 @@ nested_tensor([
   tensor([3., 4., 5., 6., 7.], device='cuda:0')
 ])
 
+Note that the passed Tensors are being copied into a contiguous piece of memory. The resulting
+NestedTensor allocates new memory to store them and does not keep a reference.
+
+In order to form a valid NestedTensor the passed Tensors also all need to match in dimension.
```
Comment (vadimkantorov) [inline docs/source/nested.rst:45]: "match in dimension" => meaning in number of dimensions? or in dimension values?

it's not very clear what this means... also not very clear currently if differently-sized images can be put inside? e.g. two tensors of shapes: [(3, 240, 320), (3, 640, 480)].

It would be very useful to explain concrete thought-of usage examples (e.g. variable-sized BTC (?) tensors) and some concrete example (images in vision) that don't work for now. I'm mentioning this e.g. because in DETR codebase a similar NestedTensor name was used to mean TensorList with two variable spatial dimensions.

File: docs/source/nested.rst [hunk 2]
```diff
@@ -39,6 +39,20 @@ nested_tensor([
...
+
+At this moment we only support one level of nesting, i.e. a simple, flat list of Tensors. In the future
```
Comment (vadimkantorov) [inline docs/source/nested.rst:47]: flat list of tensors could very well be a flat list of tensors with two unmatched dims (like images with different resolutions)

File: docs/source/nested.rst (hunk 1 above)
Comment (cpuhrsch) [reply docs/source/nested.rst:45]: Ah, good point! Thanks for the review!

I wrote a lot about this in [pytorch/nestedtensor](https://colab.research.google.com/github/pytorch/nestedtensor/blob/master/tutorials/notebooks/basic.ipynb#scrollTo=LwZM_uuaW_Cg), but not all features required for this tutorial have been ported yet. We could think about setting that as a goal. What do you think?

File: docs/source/nested.rst (hunk 2 above)
Comment (cpuhrsch) [reply docs/source/nested.rst:47]: So, aside from the dimension, semantically none of the other attributes need to match (e.g. dtype, device, numel, sizes, strides, etc.). In practice [the constructor](https://github.com/pytorch/pytorch/blob/09f79e94acce30e2002063322c9fe50d116361f1/aten/src/ATen/native/nested/NestedTensorMath.cpp#L267-L307) flattens each input, concatenates them all together and records their original sizes. This concatenated piece of memory and the sizes form the NestedTensor. For the concatenation operation to succeed the dtype and device of the input Tensors need to match, but it'd be easy to extend the constructor to detect when this isn't the case and use either the default dtype / device or require the user to pass one.

File: docs/source/nested.rst (hunk 1 above)
Comment (vadimkantorov) [reply docs/source/nested.rst:45]: Well, I'm just worried that some existing design (with goal for transformers / NLP) gets baked in, and the rest is left forgotten. So yes, I think it's best to move at least some of this thoughts at very visible public docs, so users can request this stuff...

Comment (cpuhrsch) [reply docs/source/nested.rst:45]: @vadimkantorov - I'm very keen on preventing that from happening and this is a good point. NestedTensors reach far beyond the NLP + Transformers use case.

File: docs/source/nested.rst (hunk 2 above)
Comment (vadimkantorov) [reply docs/source/nested.rst:47]: ah, I see, so many dims can match. am I right that the nested tensor always "batches" on first dimension (like torch.stack with default arg would do)? (it could have been last one or some arbitrary one in theory)

File: docs/source/nested.rst [hunk 3]
```diff
@@ -39,31 +39,78 @@ nested_tensor([
...
 
+size
++++++++++++++++++++++++++
+
+Even though a NestedTensor does not support .size(), it supports the size along a given dimension if that dimension is regular.
+
+>>> a = torch.randn(50, 128) # text 1
+>>> b = torch.randn(32, 128) # text 2
+>>> nt = torch.nested_tensor([a, b], dtype=torch.float32)
+>>> nt.size(0)
```
Comment (vadimkantorov) [inline docs/source/nested.rst:77]: also, is size function usage still recommended? at some point i had a feeling that shape is now recommended over older size

File: docs/source/nested.rst [hunk 4]
```diff
@@ -39,31 +39,78 @@ nested_tensor([
...
+>>>
+KeyboardInterrupt
+>>> nt = torch.nested_tensor([a, a], dtype=torch.float32)
+>>> nt.size(0)
+2
+>>> nt.size(1)
+20
+>>> nt.size(2)
+128
+>>> torch.stack(nt.unbind()).sizes()
```
Comment (vadimkantorov) [inline docs/source/nested.rst:98]: is it ok that this says sizes and two lines below it's size?

File: docs/source/nested.rst (hunk 2 above)
Comment (cpuhrsch) [reply docs/source/nested.rst:63]: Yes, it's always the first dimension. A future, future extension of torch.stack would automatically create a NestedTensor if passed variably sized Tensors and thus could support arbitrary dimensions. Do you have a use case in mind? I suppose we could add a `dim` argument to the `nested_tensor` constructor.

File: docs/source/nested.rst (hunk 4 above)
Comment (cpuhrsch) [reply docs/source/nested.rst:98]: Fixed!

File: docs/source/nested.rst (hunk 3 above)
Comment (vadimkantorov) [reply docs/source/nested.rst:77]: well, I meant shape[idx] of course

Comment (cpuhrsch) [reply docs/source/nested.rst:77]: The problem with `.shape` is that it returns a full `torch.Size`. While `shape[idx]` as an expression will yield a valid result, the result of `.shape` is not defined. `.size(i)` effectively provides a fusion of `.shape` + `__getitem__`.

**Comment by vadimkantorov** [comment]
from the collab page it's also not clear how exactly conv2d does work with nestedtensor? does it still do padding with zeros? for some applications padding details are important

also, for Norm modules, it would be cool if the masked out elements not affect the batch/learned statistics (and hopefully do not participate in affine params gradient)

**Comment by vadimkantorov** [comment]
one more thing: collab page shows:
```
NestedSize([
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10]),
	torch.Size([10])
])
```

can this NestedSize structure be cast/converted to a LongTensor? in other words, will torch.tensor/torch.as_tensor accept this as input? For many tasks (often with bounding box rescaling), it's useful to have unpadded image sizes as a tensor for further computation

**Comment by vadimkantorov** [comment]
another question on internal stride preservation. If I pass a permuted/not-contig tensor into nestedtensor, will the prior flatten cause another copy? or will flatten+cat stuff be performed in one go? will strides/memory_format be preserved? i could imagine both decisions (one is optimizing for being contig, another for preserving memory_format)

**Comment by cpuhrsch** [comment]
@vadimkantorov 

For conv2d and batchnorm alike, the "padded" elements don't exist. That means they do not participate in the forward or backward calculations. They do not influence the denominators of the normalization calculations and they are not used during the convolutions.

NestedSize indeed can be converted to a Tensor. Further `nested_size(i)` gives slices along it. This is useful for say sentence normalization or as you point out for bounding boxes and rescaling images. We haven't migrated this API to core yet.

On passing non-contiguous Tensors to `nested_tensor`: The resulting NestedTensor will always be contiguous. All striding information is lost and a copy is performed. We can make this an option, but it seems more worthwhile to introduce `as_nested_tensor`. You've also repeatedly asked for NestedTensors to be able to act as a view of a list of appropriate tensors, which this constructor could fulfill *if* we extend the NestedTensor metadata (and kernels!) to be able to handle this level of discontinuity. For now however there is no timeline available to implement this.

**Comment by vadimkantorov** [comment]
If we pass non-contig tensors, will there be two copies: one for flatten and second for cat? Or just one (flatten+cat fused)?
